acados/ocp_nlp/ocp_nlp_common.c

```diff
@@ -2044,162 +2117,276 @@ double ocp_nlp_evaluate_merit_fun(ocp_nlp_config *config, ocp_nlp_dims *dims,
+        }
+        else
+        {
+            alpha *= 0.7;
+        }
+    }
+
+    printf("\nalpha %f\n", alpha);
 
 #endif
 
 
+    if (opts->globalization == MERIT_BACKTRACKING)
+    {
+        // Line search version Jonathan
+        // Following Leineweber1999
+        // copy out (current iterate) to work->tmp_nlp_out
+        for (i = 0; i <= N; i++)
+            blasfeo_dveccp(nv[i], out->ux+i, 0, work->tmp_nlp_out->ux+i, 0);
+
+        for (i = 0; i < N; i++)
+            blasfeo_dveccp(nx[i+1], out->pi+i, 0, work->tmp_nlp_out->pi+i, 0);
+
+        for (i = 0; i <= N; i++)
+            blasfeo_dveccp(2*ni[i], out->lam+i, 0, work->tmp_nlp_out->lam+i, 0);
+
+            // linear update of algebraic variables using state and input sensitivity
+    //        if (i < N)
+    //        {
+    //            blasfeo_dgemv_t(nu[i]+nx[i], nz[i], alpha, mem->dzduxt+i, 0, 0, mem->qp_out->ux+i, 0, 1.0, mem->z_alg+i, 0, out->z+i, 0);
+    //        }
 
+        /* initialize (Leineweber1999 M5.1) */
+        if (mem->sqp_iter[0]==0)
+        {
+            // initialize weights
+            // equality merit weights = abs( eq multipliers )
+            for (i = 0; i < N; i++)
+            {
+                for (j=0; j<nx[i+1]; j++)
+                {
+                    tmp0 = fabs(BLASFEO_DVECEL(out->pi+i, j));
+                    BLASFEO_DVECEL(work->weight_merit_fun->pi+i, j) = tmp0;
+                }
+            }
+
+            // printf("merit fun: initialize weights lam\n");
+            for (i = 0; i <= N; i++)
+            {
+                blasfeo_dveccp(2*ni[i], out->lam+i, 0, work->weight_merit_fun->lam+i, 0);
+                // blasfeo_print_dvec(nx[i+1], work->weight_merit_fun->lam+i, 0);
+            }
+        }
+        else
+        {
+            // update weights
+            // printf("merit fun: update weights, sqp_iter = %d\n", mem->sqp_iter[0]);
+            for (i = 0; i < N; i++)
+            {
+                for(j=0; j<nx[i+1]; j++)
+                {
+                    // abs(lambda) (LW)
+                    tmp0 = fabs(BLASFEO_DVECEL(out->pi+i, j));
+                    // .5 * (abs(lambda) + sigma)
+                    tmp1 = 0.5 * (tmp0 + BLASFEO_DVECEL(work->weight_merit_fun->pi+i, j));
+                    BLASFEO_DVECEL(work->weight_merit_fun->pi+i, j) = tmp0>tmp1 ? tmp0 : tmp1;
+                }
+            }
+            for (i = 0; i <= N; i++)
+            {
+                for(j=0; j<2*ni[i]; j++)
+                {
+                    // mu (LW)
+                    tmp0 = BLASFEO_DVECEL(out->lam+i, j);
+                    // .5 * (mu + tau)
+                    tmp1 = 0.5 * (tmp0 + BLASFEO_DVECEL(work->weight_merit_fun->lam+i, j));
+                    BLASFEO_DVECEL(work->weight_merit_fun->lam+i, j) = tmp0>tmp1 ? tmp0 : tmp1;
+                }
+            }
+        }
+
+        if (1) // (mem->sqp_iter[0]!=0) // TODO: why does Leineweber do full step in first SQP iter?
```

Comment:
I don't remember exactly, but it may be related to the way the Hessian approximation is initializated in the first iteration, it may be garateed that full step works fine.

Reply:
I mean, they use BFGS a lot in MUSCOD-II